art/attacks/evasion/auto_projected_gradient_descent.py

```diff
@@ -89,258 +89,255 @@ def __init__(
-                        z_y = torch.diagonal(z_y)
+                        # def difference_logits_ratio(y_true, y_pred):
+                        #     i_y_true = tf.cast(tf.math.argmax(tf.cast(y_true, tf.int32), axis=1), tf.int32)
+                        #     i_y_pred_arg = tf.argsort(y_pred, axis=1)
+                        #     # Not completely sure if the following line is correct.
+                        #     # `i_y_pred_arg[:, -2], i_y_pred_arg[:, -1]` seems closer to the output of `loss_fn` than
+                        #     # `i_y_pred_arg[:, -1], i_y_pred_arg[:, -2]`
+                        #     i_z_i = tf.where(i_y_pred_arg[:, -1] != i_y_true[:], i_y_pred_arg[:, -2],
+                        #                      i_y_pred_arg[:, -1])
+                        #
+                        #     z_1 = tf.gather(y_pred, i_y_pred_arg[:, -1], axis=1, batch_dims=0)
+                        #     z_3 = tf.gather(y_pred, i_y_pred_arg[:, -3], axis=1, batch_dims=0)
+                        #     z_i = tf.gather(y_pred, i_z_i, axis=1, batch_dims=0)
+                        #     z_y = tf.gather(y_pred, i_y_true, axis=1, batch_dims=0)
+                        #
+                        #     z_1 = tf.linalg.diag_part(z_1)
+                        #     z_3 = tf.linalg.diag_part(z_3)
+                        #     z_i = tf.linalg.diag_part(z_i)
+                        #     z_y = tf.linalg.diag_part(z_y)
+                        #
+                        #     dlr = -(z_y - z_i) / (z_1 - z_3)
+                        #
+                        #     return tf.reduce_mean(dlr)
+                        #
+                        # def loss_fn(y_true, y_pred):
+                        #     i_y_true = np.argmax(y_true, axis=1)
+                        #     i_y_pred_arg = np.argsort(y_pred, axis=1)
+                        #     i_z_i = np.where(i_y_pred_arg[:, -1] != i_y_true[:], i_y_pred_arg[:, -1],
+                        #                      i_y_pred_arg[:, -2])
+                        #
+                        #     z_1 = y_pred[:, i_y_pred_arg[:, -1]]
+                        #     z_3 = y_pred[:, i_y_pred_arg[:, -3]]
+                        #     z_i = y_pred[:, i_z_i]
+                        #     z_y = y_pred[:, i_y_true]
+                        #
+                        #     z_1 = np.diag(z_1)
+                        #     z_3 = np.diag(z_3)
+                        #     z_i = np.diag(z_i)
+                        #     z_y = np.diag(z_y)
+                        #
+                        #     dlr = -(z_y - z_i) / (z_1 - z_3)
+                        #
+                        #     return np.mean(dlr)
+                        #
+                        # self._loss_fn = loss_fn
+                        # self._loss_object = difference_logits_ratio(y_true=estimator._labels_ph,
+                        #                                             y_pred=estimator._output)
```

Comment:
Looks like this code is replaced by the two `difference_logits_ratio` classes. Can we remove these comments here?

Reply:
Not completely, the other two `difference_logits_ratio` classes are specific for PyTorch and TensorFlow v2. The commented code here would be the class specific for TensorFlow v1, but it seems to contain a small bug as described in the exception above it and we keep the code commented until it is fix.